libraries/AP_Math/quaternion.cpp

```diff
@@ -424,34 +424,47 @@ void Quaternion::from_vector312(float roll, float pitch, float yaw)
 }
 
 // create a quaternion from its axis-angle representation
-void Quaternion::from_axis_angle(Vector3f v)
+void Quaternion::from_axis_angle(const Vector3f &v)
 {
-    const float theta = v.length();
-    if (is_zero(theta)) {
-        q1 = 1.0f;
-        q2=q3=q4=0.0f;
-        return;
+    const float theta_sq = v.length_squared();
+    float re_coeff, im_coeff;
+    if (theta_sq < radians(0.1f) * radians(0.1f)) {
```

Comment:
Pardon my naivete why choose 0.1 degrees for when to use the approximation? I assume to keep the error of the approximation below some percentage? Might be good to put that error in a comment?

Reply:
No, I'm sorry, I picked 0.1 degs somewhat arbitrarily. Other libraries seem to prefer machine epsilon for the threshold, which is waaay to small. It appears that using a threshold as big as 7 degrees / 0.1 rad is enough to keep the relative error between the approximation and the true value below the machine epsilon of single-precision float, as seen in the python code below

```
theta = np.linspace(1e-10, pi, 10000)
a = theta * (0.5 - 1.0 / 48.0 * theta ** 2)
b = np.sin(0.5 * ang)
rel_diff = np.abs((a - b) / a)
print(np.rad2deg(theta[np.argmin(rel_diff < np.finfo(np.float64.eps))])) # prints 7.056705676072013
```
The 0.1 degree threshold enforces approximation accuracy for sure, and a bit too tightly too. As I mentioned in my reply to tridge, the more often the fast branch (with approximation) kicks in, the greater is the performance boost. So I'm undecided if I should pick a looser threshold. What do you think?

Reply:
Keeping the error of the estimate below FLT_EPSILON -> 7 deg seems like the right approach from a speed / accuracy tradeoff.

-- Not for this PR --
This assumes we are only dealing with floats. However,  a few parts of the code are already using doubles and more are headed that way (optionally) then the decision of speed/accuracy will become more nuanced.

Reply:
@hendjoshsr71 I took your advice and set a looser threshold, and defined it as a macro. Albeit the *magic constant* is set to 4 deg so that the error estimate for series expansion of not just sin but also cos falls below FLT_EPSILON.

**Not for this PR**: To deal with the nuances of single/double precision math, I'd propose a templated struct similar to `std::numeric_limits`, specialized for each type to provide constants like machine epsilon or angle thresholds. You see, I'm fairly interested in this area but I sense that since Ardupilot traditionally did not focus on low-level details of numerical algorithms, there might be little interest in new contributions here, e.g. this concept of adaptive numeric limits and the double-precision / templated Quaternion class mentioned earlier.

Reply:
Makes sense about the 4 degrees.